From c4048d0ce10f5c228659516aa9f2966be77e5908 Mon Sep 17 00:00:00 2001
From: "Daniel P. Berrange" <berrange@redhat.com>
Date: Mon, 10 Dec 2012 16:59:19 +0000
Subject: [PATCH] Fix some return value checks

virDomainScreenshot and virDomainSnapshotCurrent both return
pointers, so must compare " != NULL" instead of "< 0"

Signed-off-by: Daniel P. Berrange <berrange@redhat.com>
(cherry picked from commit d6f25a7834fcad5f1ee1f8ea8f942b883086f3da)
---
 Virt.xs                | 14 ++++++++++----
 lib/Sys/Virt/Domain.pm |  4 ++--
 2 files changed, 12 insertions(+), 6 deletions(-)

diff --git a/Virt.xs b/Virt.xs
index 8f70513..1114569 100644
--- a/Virt.xs
+++ b/Virt.xs
@@ -2924,15 +2924,21 @@ open_graphics(dom, idx, fd, flags=0)
           _croak_error();
 
 
-void
+SV *
 screenshot(dom, st, screen, flags=0)
       virDomainPtr dom;
       virStreamPtr st;
       unsigned int screen;
       unsigned int flags;
-  PPCODE:
-      if (virDomainScreenshot(dom, st, screen, flags) < 0)
+ PREINIT:
+      char *mimetype;
+    CODE:
+      if (!(mimetype = virDomainScreenshot(dom, st, screen, flags)))
           _croak_error();
+      RETVAL = newSVpv(mimetype, 0);
+      free(mimetype);
+  OUTPUT:
+      RETVAL
 
 
 HV *
@@ -4284,7 +4290,7 @@ current_snapshot(dom, flags=0)
       virDomainPtr dom;
       unsigned int flags;
     CODE:
-      if ((RETVAL = virDomainSnapshotCurrent(dom, flags)) < 0)
+      if (!(RETVAL = virDomainSnapshotCurrent(dom, flags)))
           _croak_error();
   OUTPUT:
       RETVAL
diff --git a/lib/Sys/Virt/Domain.pm b/lib/Sys/Virt/Domain.pm
index ca1ed31..864335b 100644
--- a/lib/Sys/Virt/Domain.pm
+++ b/lib/Sys/Virt/Domain.pm
@@ -886,13 +886,13 @@ anoymous socket pair. The C<$flags> argument should be one of
 the constants listed at the end of this document, and defaults
 to 0.
 
-=item $dom->screenshot($st, $screen, $flags)
+=item my $mimetype = $dom->screenshot($st, $screen, $flags)
 
 Capture a screenshot of the virtual machine's monitor. The C<$screen>
 parameter controls which monitor is captured when using a multi-head
 or multi-card configuration. C<$st> must be a C<Sys::Virt::Stream>
 object from which the data can be read. C<$flags> is currently unused
-and defaults to 0.
+and defaults to 0. The mimetype of the screenshot is returned
 
 =item @vcpuinfo = $dom->get_vcpu_info()
 
